File: tree/main/samples/agent/adk/gemini_enterprise/agent_engine/samples/agent/adk/rizzcharts/java/src/main/java/com/google/a2ui/samples/rizzcharts/RizzchartsAgentExecutor.java
/*
 * Copyright 2026 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.google.a2ui.samples.rizzcharts;

import com.google.a2ui.core.schema.A2uiSchemaManager;
import com.google.adk.runner.Runner;
import com.google.adk.sessions.Session;

public class RizzchartsAgentExecutor {

  public static final String A2UI_ENABLED_KEY = "a2ui_enabled";
  public static final String A2UI_CATALOG_KEY = "a2ui_catalog";
  public static final String A2UI_EXAMPLES_KEY = "a2ui_examples";

  private final String baseUrl;
  private final Runner runner;
  private final A2uiSchemaManager schemaManager;

  public RizzchartsAgentExecutor(String baseUrl, Runner runner, A2uiSchemaManager schemaManager) {
    this.baseUrl = baseUrl;
    this.runner = runner;
    this.schemaManager = schemaManager;
  }

  // Manual implementation since A2aAgentExecutor is not available
  public Session prepareSession(String sessionId) {
    Session session =
        runner
            .sessionService()
            .getSession(sessionId, "default_user", "default_app", java.util.Optional.empty())
            .blockingGet();

    // Pre-configure the session state with A2UI components
    boolean useUi = true; // In full A2A, this checks for A2A capabilities headers.

    if (!session.state().containsKey("base_url")) {
      session.state().put("base_url", baseUrl);
    }

    if (!session.state().containsKey(A2UI_ENABLED_KEY)) {
      session.state().put(A2UI_ENABLED_KEY, useUi);
    }

    if (!session.state().containsKey(A2UI_CATALOG_KEY)) {
      if (useUi) {
        session.state().put(A2UI_CATALOG_KEY, schemaManager.getSelectedCatalog());
      } else {
        session.state().put(A2UI_CATALOG_KEY, null);
      }
    }

    if (!session.state().containsKey(A2UI_EXAMPLES_KEY)) {
      if (useUi) {
        session
            .state()
            .put(A2UI_EXAMPLES_KEY, schemaManager.loadExamples(schemaManager.getSelectedCatalog()));
      } else {
        session.state().put(A2UI_EXAMPLES_KEY, null);
      }
    }

    return session;
  }
}

File: tree/main/samples/agent/adk/gemini_enterprise/agent_engine/samples/agent/adk/rizzcharts/kotlin/src/main/kotlin/com/google/a2ui/samples/rizzcharts/RizzchartsAgentExecutor.kt
/*
 * Copyright 2026 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.google.a2ui.samples.rizzcharts

import com.fasterxml.jackson.databind.ObjectMapper
import com.google.a2ui.a2a.A2uiA2a
import com.google.a2ui.core.schema.A2uiSchemaManager
import com.google.adk.sessions.Session

class RizzchartsAgentExecutor(
  private val baseUrl: String,
  private val schemaManager: A2uiSchemaManager,
) {
  companion object {
    const val A2UI_ENABLED_KEY = "a2ui_enabled"
    const val A2UI_CATALOG_KEY = "a2ui_catalog"
    const val A2UI_EXAMPLES_KEY = "a2ui_examples"

    private val objectMapper = ObjectMapper()
  }

  fun prepareSession(session: Session, rawRequestBody: Map<*, *>) {
    @Suppress("UNCHECKED_CAST")
    val requestBody = objectMapper.convertValue(rawRequestBody, Map::class.java) as Map<String, Any>
    val params = requestBody["params"] as? Map<*, *>
    val messageMap = params?.get("message") as? Map<*, *>
    val messageMetadata = messageMap?.get("metadata") as? Map<*, *>
    val rootMetadata = requestBody["metadata"] as? Map<*, *>
    val metadata = messageMetadata ?: rootMetadata
    val capabilitiesNode = metadata?.get("a2uiClientCapabilities")
    var capabilities = capabilitiesNode as? Map<*, *>

    if (capabilities == null) {
      val history = params?.get("history") as? List<*>
      if (history != null) {
        for (msg in history) {
          val msgMap = msg as? Map<*, *>
          val msgMeta = msgMap?.get("metadata") as? Map<*, *>
          val msgCaps = msgMeta?.get("a2uiClientCapabilities") as? Map<*, *>
          if (msgCaps != null) {
            capabilities = msgCaps
            break
          }
        }
      }
    }

    val requestedExtensions =
      ((params?.get("extensions") ?: requestBody["extensions"]) as? List<*>)?.filterIsInstance<
        String
      >() ?: emptyList()

    var useUi = A2uiA2a.tryActivateA2uiExtension(requestedExtensions) { /* no-op */ }
    if (!useUi) {
      useUi = capabilitiesNode != null || capabilities != null
    }

    if (!session.state().containsKey("base_url")) {
      session.state()["base_url"] = baseUrl
    }

    session.state()[A2UI_ENABLED_KEY] = useUi

    if (useUi) {
      session.state()[A2UI_CATALOG_KEY] = schemaManager.getSelectedCatalog()
      session.state()[A2UI_EXAMPLES_KEY] =
        schemaManager.loadExamples(schemaManager.getSelectedCatalog())
    } else {
      session.state().remove(A2UI_CATALOG_KEY)
      session.state().remove(A2UI_EXAMPLES_KEY)
    }
  }
}
